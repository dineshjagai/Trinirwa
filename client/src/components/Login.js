/* eslint-disable no-unused-vars */
import React, { useEffect, useState } from 'react';
import Axios from 'axios';
import {
  Link, useHistory, BrowserRouter as Route,
} from 'react-router-dom';
import '../App.css';
import { userLogin, getUid } from './Module';
import Home from './Home';

export default function Login() {
  const [username, setUsername] = useState('');
  const [password, setPassword] = useState('');
  const history = useHistory();
  const [uid, setUid] = useState('');
  const [count, setCount] = useState(0);
  const [isAuth, setIsAuth] = useState(false);
  Axios.defaults.withCredentials = true;

  useEffect(() => setUid(uid), [uid, username, password, count, isAuth]);
  // useEffect(() => console.log(count), [count]);

  const login = async () => {
    userLogin(username, password).then((response) => {
      if (response.data.message) {
        alert(response.data.message);
        // window.history.replaceState(null, 'new page', '/login');
        // window.location.reload();
        setUsername('');
        setPassword('');
        setUid('');
        setCount(count + 1);
        if (count >= 3) {
          // disable user from logging into the browser
          // TODO implement this
        }
        document.getElementById('new-password-form').value = '';
        document.getElementById('new-username-form').value = '';
      } else {
        setIsAuth(true);
        let fetchedUid = null;
        getUid(username).then((res) => {
          [fetchedUid] = [Array.from(res.data.data)[0].uid];
          setUid(fetchedUid);
          // window.history.replaceState({ fetchedUid }, 'new page', '/home');
          // window.location.reload();
        }).catch((e) => {
          console.log(e);
        });
      }
    });
  };

  return (
    <div className="card shadow mx-auto mt-5" style={{ width: '30rem' }}>
      <div className="container mx-1">
        <div className="card-body p-xs">
          <h2 className="card-title text-center font-weight-bold">Log In</h2>
          <div className="form-group">
            <label>Username:</label>
            <input
              id="new-username-form"
              placeholder="Enter Username"
              className="form-control"
              onChange={(e) => setUsername(e.target.value)}
            />
          </div>
          <div className="form-group">
            <label>Password:</label>
            <input
              id="new-password-form"
              type="password"
              placeholder="Enter Password"
              className="form-control"
              onChange={(e) => setPassword(e.target.value)}
            />
          </div>
          <p>
            Forgot Password?&nbsp;
            <Link to="/">Click here to reset!</Link>
          </p>
          <div className="form-group text-center">
            {isAuth ? (
              <Link to="/home">
                <button
                  type="button"
                  className="btn btn-primary w-100"
                  onClick={login}
                >
                  Log In
                </button>
              </Link>
            ) : (
              <button
                type="button"
                className="btn btn-primary w-100"
                onClick={login}
              >
                Log In
              </button>
            )}
          </div>
          <p>
            Don&apos;t have an account?&nbsp;
            <Link to="/registration">Sign Up!</Link>
            <Route path="/home/" component={() => <Home uid={uid} />} />
          </p>
        </div>
      </div>
    </div>
  );
}
