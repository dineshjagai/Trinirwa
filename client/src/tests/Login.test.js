import React from 'react';
import renderer from 'react-test-renderer';
import Login from '../components/Login';
import { render, screen, fireEvent } from '@testing-library/react'
import { act } from 'react-dom/test-utils'

describe("Login testing", () => {
    test('Login Snapshot', () => {
        const component = renderer.create(<Login />);
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('checks that the labels are there', () => {
        const { getByText } = render(<Login />);
        getByText('Username:');
        getByText('Password:');
    });

    test('calls Log In button', () => {
        const handleClick = jest.fn();
        render(<Login />);
        fireEvent.click(screen.getByText('Log In', { selector: 'button' }));
        expect(handleClick).toHaveBeenCalledTimes(0)
    });

    test("with valid inputs", async () => {
        const { getByPlaceholderText, getByRole } = render(<Login />);
        const usernameLabel = getByPlaceholderText('Enter Username');
        const passwordLabel = getByPlaceholderText('Enter Password');      
        const mockOnSubmit = jest.fn();

        await act(async () => {
            fireEvent.change(usernameLabel, {target: {value: "isimbi"}});
            fireEvent.change(passwordLabel, {target: {value: "Anaick@123"}});
        })

        await act(async () => {
            fireEvent.click(getByRole("button"));
        })

      expect(mockOnSubmit).toHaveBeenCalledTimes(0);
    });

    test("with invalid username", () => {

    });

   test("with invalid password", () => {

    });

});