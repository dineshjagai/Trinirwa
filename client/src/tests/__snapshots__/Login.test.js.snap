// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login testing Login Snapshot 1`] = `
<div
  className="card shadow mx-auto mt-5"
  style={
    Object {
      "width": "30rem",
    }
  }
>
  <div
    className="container mx-1"
  >
    <div
      className="card-body p-xs"
    >
      <h2
        className="card-title text-center font-weight-bold"
      >
        Log In
      </h2>
      <div
        className="form-group"
      >
        <label>
          Username:
        </label>
        <input
          className="form-control"
          id="new-username-form"
          onChange={[Function]}
          placeholder="Enter Username"
        />
      </div>
      <div
        className="form-group"
      >
        <label>
          Password:
        </label>
        <input
          className="form-control"
          id="new-password-form"
          onChange={[Function]}
          placeholder="Enter Password"
          type="password"
        />
      </div>
      <p>
        Forgot Password? 
        <a
          href="/resetPassword"
          onClick={[Function]}
        >
          Click here to reset!
        </a>
      </p>
      <div
        className="form-group text-center"
      >
        <button
          className="btn btn-primary w-100"
          onClick={[Function]}
          type="submit"
        >
          Log In
        </button>
      </div>
      <p>
        Don't have an account? 
        <a
          href="/registration"
          onClick={[Function]}
        >
          Sign Up!
        </a>
      </p>
    </div>
  </div>
</div>
`;
